keras/backend/tensorflow_backend.py

```diff
@@ -193,6 +201,64 @@ def set_session(session):
     _SESSION = session
 
 
+# DEVICE MANIPULATION AND PROBING
+
+class _TfDeviceCaptureOp(object):
+    """Class for capturing the tf device scope."""
+
+    def __init__(self):
+        """__init__ method."""
+        self.device = None
+
+    def _set_device(self, device):
+        """_set_device method."""
+        self.device = device
+
+
+def get_current_tf_device():
+    """Return device string of current graph context that's explicitly, otherwise returns `None`."""
+    g = tf.get_default_graph()
+    op = _TfDeviceCaptureOp()
+    g._apply_device_functions(op)
+    return op.device
+
+
+def is_current_explicit_device(device_type):
+    """
+    Check if the current device is explicitly set on the device type.
+
+    Args:
+        device_type (str): 'GPU' or 'CPU'
+    Returns:
+        bool: if the current device scope is explicitly set on the device type.
+    """
+    device_type = device_type.upper()
+    if device_type not in ['CPU', 'GPU']:
+        raise ValueError('device_type should be either "CPU" or "GPU".')
+    device = get_current_tf_device()
+    return (device is not None and device.device_type == device_type.upper())
+
+
+def get_available_gpus():
+    """Get a list of available gpu devices (formatted as strings)."""
+    return [x.name for x in LOCAL_DEVICES if x.device_type == 'GPU']
+
+
+def has_nchw_support():
```

Comment:
This sounds too specific. Why not something like "_running_on_gpu()" or something to that extent?

Reply:
Hmm.. What it checks for is if the current scope is "not explicit on CPU, and has GPUs available". It's called like this to anticipate any TF 1.x's support that does not need the roundtrip so we can test for the tf version in the function.